apollo-router-core/src/services/mod.rs

```diff
@@ -206,4 +178,58 @@ impl<L> ServiceBuilderExt<L> for ServiceBuilder<L> {
             },
         )
     }
+
+    fn dedup<S, Request, Key>(
+        self,
+        key_fn: fn(&Request) -> Option<&Key>,
+        request_fn: fn(&Key) -> Request,
+        merge_fn: fn(Request, S::Response) -> S::Response,
+    ) -> ServiceBuilder<Stack<DeduplicationLayer<S, Request, Key>, L>>
+    where
+        Request: Send,
+        Key: Clone + Eq + Hash + Send,
+        S: Service<Request> + Send + Clone,
+        <S as Service<Request>>::Response: Clone + Send,
+        <S as Service<Request>>::Error: Into<BoxError> + Send + Sync,
+        <S as Service<Request>>::Future: Send,
+    {
+        self.layer(DeduplicationLayer::new(key_fn, request_fn, merge_fn))
+    }
+
+    fn dedup_subgraph_requests<S>(
+        self,
+    ) -> ServiceBuilder<
+        Stack<DeduplicationLayer<S, SubgraphRequest, http_compat::Request<Request>>, L>,
+    >
+    where
+        S: Service<SubgraphRequest, Response = SubgraphResponse> + Send + Clone,
+        <S as Service<SubgraphRequest>>::Response: Clone + Send,
+        <S as Service<SubgraphRequest>>::Error: Into<BoxError> + Send + Sync,
+        <S as Service<SubgraphRequest>>::Future: Send,
+    {
+        self.dedup(
+            |request: &SubgraphRequest| match request.operation_kind {
+                OperationKind::Query => Some(&request.http_request),
```

Comment:
we will need to make this configurable soon: if an Authorization header is passed from the client to the backend, it should be part of the hashed key and the subgraph request.

Are there cases where the set of headers we want to hash and the set of headers we want to send to the subgraph are different?

Reply:
Agree, and there would definitely be a case for disabling completely.
However this should be tackled as a separate ticket.

Reply:
so there's one case I can think of, for https://github.com/apollographql/router/issues/517.
Right now the otel headers are generated directly by reqwest-middleware, taking them from some global context generated in warp_http_server_factory (separate from the `Context` used in extensions.

For #517, we need to make the current span id (and possibly the trace id) available to a plugin that will add a header with the right format to subgraph requests. If that plugin applies before the dedup, we will lose the opentracing headers, while it would make sense to keep them for the only request actually sent. If the plugin runs after deduplication, it should be fine.

Filtering the request to create the deduplicated request makes sense, but maybe it should not be created from the key, but from the real request instead?

Reply:
There is a good case for making this a plugin that is not enabled by default.
1. There is potential that it will cause incorrect results. Graphql queries are not guaranteed to be a function of the requestv e.g: https://graphql.org/graphql-js/object-types/
2. It's going to subvert our perf testing. If we introduce an issue which makes our subgraph fetching slower we're not going to see it.